sympy/polys/ring_series.py

```diff
@@ -490,9 +622,363 @@ def rs_exp(p, x, prec):
     r = rs_series_from_list(p, c, x, prec)
     return r
 
+# TODO
+# Needs to be benchmarked before use in rs_atan
+def _atan_series(p, iv, prec):
+    ring = p.ring
+    mo = ring(-1)
+    c = [-mo]
+    p2 = rs_square(p, iv, prec)
+    for k in range(1, prec):
+        c.append(mo**k/(2*k + 1))
+    s = rs_series_from_list(p2, c, iv, prec)
+    s = rs_mul(s, p, iv, prec)
+    return s
+
+def rs_atan(p, x, prec):
+    """
+    The arctangent of a series
+
+    Returns the series expansion of the atan of p, about 0.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.domains import QQ
+    >>> from sympy.polys.rings import ring
+    >>> from sympy.polys.ring_series import rs_atan
+    >>> R, x, y = ring('x, y', QQ)
+    >>> rs_atan(x + x*y, x, 4)
+    -1/3*x**3*y**3 - x**3*y**2 - x**3*y - 1/3*x**3 + x*y + x
+
+    See Also
+    ========
+
+    atan
+    """
+    if _has_constant_term(p, x):
+        raise NotImplementedError('Polynomial must not have constant term in \
+              the series variables')
+    ring = p.ring
+
+    # Instead of using a closed form formula, we differentiate atan(p) to get
+    # `1/(1+p**2) * dp`, whose series expansion is much easier to calculate.
+    # Finally we integrate to get back atan
+    dp = p.diff(x)
+    p1 = rs_square(p, x, prec) + ring(1)
+    p1 = rs_series_inversion(p1, x, prec - 1)
+    p1 = rs_mul(dp, p1, x, prec - 1)
+    return rs_integrate(p1, x)
+
+def _tan1(p, x, prec):
```

Comment:
@certik My previous comment about `_tan1` is not visible due to new commits. I am pasting it again:
This function uses Newton's method to approximate tan. Say I want to calculate tan(p), I write it as x = tan(p) (assuming x to be the series expansion) => p - atan(x) = 0. Now I use Newton iterations to solve for x. I had assumed that this will work only for univariate p, but it seems to work even for multivariate p.
```
In [12]: (tan(a + a*b).series(a, 0, 5)).expand()
Out[12]: a + a*b + a**3/3 + a**3*b + a**3*b**2 + a**3*b**3/3 + O(a**5)
In [13]: _tan1(x + x*y, x, 5)
Out[13]: 1/3*x**3*y**3 + x**3*y**2 + x**3*y + 1/3*x**3 + x*y + x
```
If so, I think we can do away with the call to fun.